### PR TITLE
More hooks for Scripts/Modules

### PR DESCRIPTION
**Pull request by 3ndos**
**Changes proposed:**

-  Adds more hooks for Scripts/Modules specifically for https://github.com/3ndos/VASAutoBalanceModule , https://github.com/3ndos/RandomEnchantsModule and
https://github.com/3ndos/AnticheatModule

**Target branch(es):** 1.x/2.x etc.

**Issues addressed:** Closes #

**Tests performed:** (Does it build, tested in-game, etc) Builds fine, Tested in-game with modules linked above.

**Known issues and TODO list:**

- Comments were made for me, feel free to remove them when merging  
- I need to double-check indentations, passed this between deb8&windows so god only knows

**NOTE** You no longer need to squash your commits, on merge we will squash it for you. (GitHub added a new feature)

First time using GIT so any tips would be great! Thanks.

### CODE REVIEW

**Comment by ShinDarth** [comment]
@3ndos thanks for this PR, can you please remove all the comments that surround the hooks ? we should try to keep the code as cleaner as possible. I know credits are important, but every committer will be automatically listed in our [contributors list](https://github.com/azerothcore/azerothcore-wotlk/graphs/contributors)

File: src/game/Scripting/ScriptMgr.cpp [hunk 1]
```diff
@@ -601,7 +601,9 @@ void ScriptMgr::OnPlayerEnterMap(Map* map, Player* player)
 {
     ASSERT(map);
     ASSERT(player);
-
+	/*Additions for VAS_AutoBalance*/
+	FOREACH_SCRIPT(AllMapScript)->OnPlayerEnterAll(map, player);
+	/*End of Additions for VAS_AutoBalance*/
```
Comment (ShinDarth) [inline src/game/Scripting/ScriptMgr.cpp]: indentation

File: src/game/Scripting/ScriptMgr.cpp [hunk 2]
```diff
@@ -621,7 +623,9 @@ void ScriptMgr::OnPlayerLeaveMap(Map* map, Player* player)
 {
     ASSERT(map);
     ASSERT(player);
-
+	/*Additions for VAS_AutoBalance*/
+	FOREACH_SCRIPT(AllMapScript)->OnPlayerLeaveAll(map, player);
+	/*End of Additions for VAS_AutoBalance*/
```
Comment (ShinDarth) [inline src/game/Scripting/ScriptMgr.cpp]: indentation

File: src/game/Scripting/ScriptMgr.cpp [hunk 3]
```diff
@@ -815,6 +819,10 @@ void ScriptMgr::OnCreatureUpdate(Creature* creature, uint32 diff)
 {
     ASSERT(creature);
 
+	/*Additions for VAS_AutoBalance*/
+	FOREACH_SCRIPT(AllCreatureScript)->OnAllCreatureUpdate(creature, diff);
+	/*End of Additions for VAS_AutoBalance*/
+
```
Comment (ShinDarth) [inline src/game/Scripting/ScriptMgr.cpp]: indentation

File: src/game/Scripting/ScriptMgr.cpp [hunk 4]
```diff
@@ -1454,7 +1476,65 @@ void ScriptMgr::OnBeforeUpdateArenaPoints(ArenaTeam* at, std::map<uint32, uint32
...
+{
+	FOREACH_SCRIPT(UnitScript)->ModifySpellDamageTaken(target, attacker, damage);
+}
 
+void ScriptMgr::ModifyHealRecieved(Unit* target, Unit* attacker, uint32& damage)
+{
+	FOREACH_SCRIPT(UnitScript)->ModifyHealRecieved(target, attacker, damage);
+}
+AllMapScript::AllMapScript(const char* name)
+	: ScriptObject(name)
```
Comment (ShinDarth) [inline src/game/Scripting/ScriptMgr.cpp]: maybe it's better all on the same line, it's not too long after all :D

File: src/game/Scripting/ScriptMgr.h [hunk 5]
```diff
@@ -288,7 +288,6 @@ class FormulaScript : public ScriptObject
         // Called when calculating the experience rate for group experience.
         virtual void OnGroupRateCalculation(float& /*rate*/, uint32 /*count*/, bool /*isRaid*/) { }
 };
-
```
Comment (ShinDarth) [inline src/game/Scripting/ScriptMgr.h]: why removing this line? :D

File: src/game/Spells/Auras/SpellAuraEffects.cpp [hunk 6]
```diff
@@ -6070,7 +6070,10 @@ void AuraEffect::HandlePeriodicDamageAurasTick(Unit* target, Unit* caster) const
 
     // ignore non positive values (can be result apply spellmods to aura damage
     uint32 damage = std::max(GetAmount(), 0);
-
+	/*Additions for VAS_AutoBalance*/
+	// Script Hook For HandlePeriodicDamageAurasTick -- Allow scripts to change the Damage pre class mitigation calculations
+	sScriptMgr->ModifyPeriodicDamageAurasTick(target, caster, damage);
+	/*End of Additions for VAS_AutoBalance*/
```
Comment (ShinDarth) [inline src/game/Spells/Auras/SpellAuraEffects.cpp]: indentation

File: src/game/Spells/Auras/SpellAuraEffects.cpp [hunk 7]
```diff
@@ -6183,7 +6186,11 @@ void AuraEffect::HandlePeriodicHealthLeechAuraTick(Unit* target, Unit* caster) c
     CleanDamage cleanDamage = CleanDamage(0, 0, BASE_ATTACK, MELEE_HIT_NORMAL);
 
     uint32 damage = std::max(GetAmount(), 0);
-    if (GetBase()->GetType() == DYNOBJ_AURA_TYPE)
+	/*Additions for VAS_AutoBalance*/
+	// Script Hook For HandlePeriodicHealthLeechAurasTick -- Allow scripts to change the Damage pre class mitigation calculations
+	sScriptMgr->ModifyPeriodicDamageAurasTick(target, caster, damage);
+	/*End of Additions for VAS_AutoBalance*/
+	if (GetBase()->GetType() == DYNOBJ_AURA_TYPE)
```
Comment (ShinDarth) [inline src/game/Spells/Auras/SpellAuraEffects.cpp]: indentation

File: src/game/Spells/Auras/SpellAuraEffects.cpp [hunk 8]
```diff
@@ -6370,6 +6377,11 @@ void AuraEffect::HandlePeriodicHealAurasTick(Unit* target, Unit* caster) const
 
     uint32 absorb = 0;
     uint32 heal = uint32(damage);
+	/*Additions for VAS_AutoBalance*/
+	//Tried to decipher this jumbled mess, but this will have to do -- 3ndos
+	// Script Hook For HandlePeriodicDamageAurasTick -- Allow scripts to change the Damage pre class mitigation calculations
+	sScriptMgr->ModifyPeriodicDamageAurasTick(target, caster, heal);
```
Comment (ShinDarth) [inline src/game/Spells/Auras/SpellAuraEffects.cpp]: indentation

File: src/game/Entities/Unit/Unit.cpp [hunk 9]
```diff
@@ -10209,8 +10221,10 @@ int32 Unit::HealBySpell(Unit* victim, SpellInfo const* spellInfo, uint32 addHeal
     uint32 absorb = 0;
     // calculate heal absorb and reduce healing
     CalcHealAbsorb(victim, spellInfo, addHealth, absorb);
-
-    int32 gain = Unit::DealHeal(this, victim, addHealth);
+    
+	sScriptMgr->ModifyHealRecieved(this, victim, addHealth);
+    
+	int32 gain = Unit::DealHeal(this, victim, addHealth);
```
Comment (ShinDarth) [inline src/game/Entities/Unit/Unit.cpp]: indentation

**Comment by ShinDarth** [comment]
@3ndos let me know when this is ready

File: src/game/Entities/Item/Item.cpp [hunk 10]
```diff
@@ -652,8 +655,8 @@ void Item::SetState(ItemUpdateState state, Player* forplayer)
         // new items must stay in new state until saved
         if (uState != ITEM_NEW)
             uState = state;
-
-        AddToUpdateQueueOf(forplayer);
+        if(forplayer)
+            AddToUpdateQueueOf(forplayer);
```
Comment (ShinDarth) [inline src/game/Entities/Item/Item.cpp]: missing space after if

File: src/game/Entities/Item/Item.cpp [hunk 11]
```diff
@@ -642,8 +642,11 @@ void Item::SetState(ItemUpdateState state, Player* forplayer)
     if (uState == ITEM_NEW && state == ITEM_REMOVED)
     {
         // pretend the item never existed
-        RemoveFromUpdateQueueOf(forplayer);
-        forplayer->DeleteRefundReference(GetGUIDLow());
+        if(forplayer)
```
Comment (ShinDarth) [inline src/game/Entities/Item/Item.cpp]: missing space after if

**Comment by falzard** [comment]
When is this going to get accepted to master? (Not a demand, just a question)

**Comment by ShinDarth** [comment]
@falzard I have no problem merging this, just let me know when it's done.

For future, use [WIP] tag in PR's title as long as it is a Work In Progress, so when you finish and remove the tag we know it's done

**Comment by 3ndos** [comment]
It's done.

**Pull request merged**